crates/rpc/eth/logs.rs

```diff
@@ -0,0 +1,181 @@
+    fn handle(&self, storage: Store) -> Result<Value, RpcErr> {
+        let Ok(Some(from)) = self.from_block.resolve_block_number(&storage) else {
+            return Err(RpcErr::BadParams);
+        };
+        let Ok(Some(to)) = self.to_block.resolve_block_number(&storage) else {
+            return Err(RpcErr::BadParams);
+        };
+
+        let address_filter: BTreeSet<_> = match &self.address_filters {
+            Some(AddressFilter::Single(address)) => std::iter::once(address).collect(),
+            Some(AddressFilter::Many(addresses)) => addresses.iter().collect(),
+            None => BTreeSet::new(),
+        };
+
+        let mut logs: Vec<RpcLog> = Vec::new();
+        // The idea here is to fetch every log and filter by address, if given.
+        // For that, we'll need each block in range, and its transactions,
+        // and for each transaction, we'll need its receipts, which
+        // contain the actual logs we want.
+        for block_num in from..=to {
+            // Take the header of the block, we
+            // will use it to access the transactions.
+            let block_body = storage.get_block_body(block_num)?.ok_or(RpcErr::Internal)?;
+            let block_header = storage
+                .get_block_header(block_num)?
+                .ok_or(RpcErr::Internal)?;
+            let block_hash = block_header.compute_block_hash();
+
+            let mut block_log_index = 0_u64;
+
+            // Since transactions share indices with their receipts,
+            // we'll use them to fetch their receipts, which have the actual logs.
+            for (tx_index, tx) in block_body.transactions.iter().enumerate() {
+                let tx_hash = tx.compute_hash();
+                let receipt = storage
+                    .get_receipt(block_num, tx_index as u64)?
+                    .ok_or(RpcErr::Internal)?;
+
+                if receipt.succeeded {
+                    for log in &receipt.logs {
+                        if address_filter.is_empty() || address_filter.contains(&log.address) {
+                            // Some extra data is needed when
+                            // forming the RPC response.
+                            logs.push(RpcLog {
+                                log: log.clone().into(),
+                                log_index: block_log_index,
+                                transaction_hash: tx_hash,
+                                transaction_index: tx_index as u64,
+                                block_number: block_num,
+                                block_hash,
+                                removed: false,
+                            });
+                        }
+                        block_log_index += 1;
+                    }
+                }
+            }
+        }
+        // Now that we have the logs filtered by address,
+        // we still need to filter by topics if it was a given parameter.
+
+        let filtered_logs = logs
```

Comment:
should you have a `if self.topics.len()` before starting this loop?

Reply:
I've added one, and the tests still pass, so yeah.